### ISSUE TITLE
We should document that `giterr_last()` should not be called except after an error

### ISSUE THREAD

**Issue by KizzyCode** [libgit2/libgit2#4772]
### Reproduction steps
 1. Create a new repository
 2. Create an initial commit
 3. Check `giterr_last()` which returns an error even if everything was successful

```c
// Init libgit
assert(git_libgit2_init() == 1);

// Create a new repo
git_repository* repo;
assert(git_repository_init(&repo, "empty", 0) == 0);

// Get index
git_index* index;
assert(git_repository_index(&index, repo) == 0);

// Get tree-ID and tree
git_oid tree_id;
git_index_write_tree(&tree_id, index);

git_tree* tree;
assert(git_tree_lookup(&tree, repo, &tree_id) == 0);

// Create signature
git_signature* signature;
assert(git_signature_now(&signature, "Max Musterfrau", "max@musterfrau.invalid") == 0);

// Check for error
assert(giterr_last() == NULL);

// Create commit
git_oid commit_id;
assert(git_commit_create(
    &commit_id, repo, "HEAD",
    signature, signature, NULL, "Initial commit",
    tree, 0, NULL
) == 0); // <- This works

// Check for error
git_error const* error = giterr_last();
if (error != NULL) printf("Git-error: %s\n", error->message); // <- This prints "Git-error: reference 'refs/heads/master' not found" even if everything was successful
```

### Expected behavior
`giterr_last()` should return `NULL`

### Actual behavior
`giterr_last()` returns an error even if everything was successful

### Version of libgit2 (release number or SHA1)
0.27.4 and c1ef58a

### Operating system(s) tested
macOS High Sierra (17G65)

**Comment by ethomson** [comment]
It shouldn’t, in fact.  If you didn’t get an error return code, you should not check the error string.

There are various errors that occur at low layers of the application (eg file not found) that we ignore and do not bother to clear.

I don’t have a strong interest in changing this behavior, as there’s no benefit, and we will _certainly_ miss cases.

**Comment by KizzyCode** [comment]
Ok, I see your point. Maybe it's a good idea to add this to the documentation?

**Comment by ethomson** [comment]
> Ok, I see your point. Maybe it's a good idea to add this info to the documentation?

Yes you are 100% correct.  Going to reopen this so we can ensure that we do that.

**Comment by ethomson** [comment]
Fixed via #4773.  Thanks for reporting this, and thanks for fixing it @RandomSort.
